### PR TITLE
[GPU] Added more information for memory allocation in DumpProfilingData 

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
 - Added more information for memory allocation in DumpProfilingData (new_alloc/reuse_buffer/can_be_optimized/from_pool)

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Review by e-ddykim — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 1]
```diff
@@ -672,23 +682,30 @@ event::ptr primitive_inst::realloc_if_needed() {
...
                 } else {
                     // i-th layout has not been allocated yet
-                    _intermediates_memory.push_back(allocate_internal_buffer(i, need_reset));
+                    _intermediates_memory.push_back(allocate_internal_buffer(i, need_reset, &from_pool));
                     max_intermediates_memory_sizes.push_back(_intermediates_memory[i]->size());
                 }
+                if (from_pool) {
+                    std::string mem_info = "_ibuf" + std::to_string(i) + ":from_pool";
+                } else {
+                    std::string mem_info = "_ibuf" + std::to_string(i) + ":from_pool";
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:707]: It looks like `:new_alloc` was intended here

**Review by andrew-k-park — APPROVED** [review]
Overall LTGM except for one minor thing

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Fixed! thank you for checking!

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp [hunk 2]
```diff
@@ -107,19 +107,22 @@ class memory_pool {
                           const std::set<primitive_id>& restrictions,
                           allocation_type type,
                           bool reusable = true,
-                          bool reset = true);  // get from pool or create memory allocation
+                          bool reset = true,
+                          bool* from_pool = nullptr);  // get from pool or create memory allocation
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp:111]: Wouldn't it be better to introduce some new field for memory structure? Something like this:
```
struct memory {   
    ...
    std::shared_ptr<MemoryTracker> get_mem_tracker() const { return m_mem_tracker; }

    // Debug memory allocation info
    GPU_DEBUG_CODE(std::string mem_alloc_info);

protected:
    ...
};
```
And set it in `memory_pool::get_from_non_padded_pool` functions as follow:

```
                ...
                auto ret_mem = _engine->reinterpret_buffer(*it->second._memory, layout);
                GPU_DEBUG_CODE(ret_mem->mem_alloc_info = "from_pool");
                return ret_mem;
            }
        }
        auto mem = alloc_memory(layout, type);
        GPU_DEBUG_CODE(mem->mem_alloc_info = "new_alloc");
```
And use this field for statistic collection, just to minimize passing this `from_pool` all over functions call:
`GPU_DEBUG_PROFILED_STAGE_MEMALLOC_INFO(memory->mem_alloc_info);`

Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp:111]: Thats a good suggestion, I will apply it. Thanks :)

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 3]
```diff
@@ -672,23 +682,30 @@ event::ptr primitive_inst::realloc_if_needed() {
         const auto& ibuf_layouts = _impl->get_internal_buffer_layouts();
         if (ibuf_layouts.empty())
             return ev;
-
         for (size_t i = 0; i < ibuf_layouts.size(); ++i) {
             if (i < _intermediates_memory.size() && ibuf_layouts[i].bytes_count() <= max_intermediates_memory_sizes[i]) {
                 // can reuse
                 _intermediates_memory[i] = _network.get_engine().reinterpret_buffer(*_intermediates_memory[i], ibuf_layouts[i]);
+                std::string mem_info = "_ibuf" + std::to_string(i) + ":new_alloc";
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: this code is removed :) thanks

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 4]
```diff
@@ -672,23 +682,30 @@ event::ptr primitive_inst::realloc_if_needed() {
...
                     // i-th layout has not been allocated yet
-                    _intermediates_memory.push_back(allocate_internal_buffer(i, need_reset));
+                    _intermediates_memory.push_back(allocate_internal_buffer(i, need_reset, &from_pool));
                     max_intermediates_memory_sizes.push_back(_intermediates_memory[i]->size());
                 }
+                if (from_pool) {
+                    std::string mem_info = "_ibuf" + std::to_string(i) + ":from_pool";
+                } else {
+                    std::string mem_info = "_ibuf" + std::to_string(i) + ":new_alloc";
+                }
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: It was fixed, thanks.,

File: src/plugins/intel_gpu/src/runtime/memory_pool.cpp [hunk 5]
```diff
@@ -168,6 +176,8 @@ memory::ptr memory_pool::get_from_padded_pool(const layout& layout,
                 !has_conflict(rec_list._users, restrictions, network_id)) {
                 rec_list._users.insert({id, network_id});
                 auto ret_mem = _engine->reinterpret_buffer(*(rec_list._memory), layout);
+                if (from_pool)
+                    *from_pool = false;
```
[the remark this answers is not in the document]
Comment (yeonbok) [reply src/plugins/intel_gpu/src/runtime/memory_pool.cpp]: It was fixed, thanks

File: src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/memory_pool.hpp]: I updated PR. Please review!

**Review by sshlyapn — APPROVED** [review]

**Pull request merged**